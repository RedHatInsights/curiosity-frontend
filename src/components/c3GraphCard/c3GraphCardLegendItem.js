import React from 'react';
import PropTypes from 'prop-types';
import { Button, Tooltip, TooltipPosition } from '@patternfly/react-core';
import { EyeSlashIcon } from '@patternfly/react-icons';
import { helpers } from '../../common';

/**
 * A chart/graph legend button.
 *
 * @augments React.Component
 * @fires onFocus
 * @fires onRevert
 * @fires onToggle
 */
class C3GraphCardLegendItem extends React.Component {
  state = { updatedIsToggled: null };

  /**
   * C3 chart display focus.
   *
   * @event onFocus
   */
  onFocus = () => {
    const { updatedIsToggled } = this.state;
    const { chart, chartId, getToggle, isToggled } = this.props;

    if ((typeof getToggle === 'function' && !isToggled) || !updatedIsToggled) {
      chart.focus(chartId);
    }
  };

  /**
   * C3 chart display revert.
   *
   * @event onRevert
   */
  onRevert = () => {
    const { updatedIsToggled } = this.state;
    const { chart, getToggle, isToggled } = this.props;

    if ((typeof getToggle === 'function' && !isToggled) || !updatedIsToggled) {
      chart.revert();
    }
  };

  /**
   * C3 chart display toggle.
   *
   * @event onToggle
   */
  onToggle = () => {
    const { updatedIsToggled } = this.state;
    const { chart, chartId, getToggle, isToggled } = this.props;

    chart.toggle(chartId);

    if (typeof getToggle === 'function') {
      getToggle({ isToggled: !isToggled });
    } else {
      this.setState({
        updatedIsToggled: !updatedIsToggled
      });
    }

    this.onRevert();
  };

  /**
   * C3 chart display config color.
   *
   * @returns {string}
   */
  getColor = () => {
    const { chart, chartId } = this.props;
    return chart.color(chartId);
  };

  /**
   * ToDO: evaluate using nullish coalescing operator for checkIsToggled
   */
  /**
   * Render a chart legend item.
   *
   * @returns {Node}
   */
  render() {
    const { updatedIsToggled } = this.state;
    const { children, chart, chartId, isDisabled, isThreshold, isToggled, tooltipContent } = this.props;
    const checkIsToggled =
      (updatedIsToggled === null && typeof isToggled === 'boolean' && isToggled) || updatedIsToggled || false;

    if (checkIsToggled) {
      chart.hide(chartId);
    }

    const button = (
      <Button
        tabIndex={0}
        key={`curiosity-button-${chartId}`}
        variant="link"
        onClick={this.onToggle}
        onFocus={this.onFocus}
        onMouseOver={this.onFocus}
        onBlur={this.onRevert}
        onMouseOut={this.onRevert}
        onKeyPress={this.onToggle}
        component="a"
        isDisabled={isDisabled}
        icon={
          ((isDisabled || checkIsToggled) && <EyeSlashIcon />) ||
          (isThreshold && (
            <hr
              aria-hidden
              className="threshold-legend-icon"
              style={{
                visibility: (isDisabled && 'hidden') || (checkIsToggled && 'hidden') || 'visible',
                borderTopColor: this.getColor()
              }}
            />
          )) || (
            <div
              aria-hidden
              className="legend-icon"
              style={{
                visibility: (isDisabled && 'hidden') || (checkIsToggled && 'hidden') || 'visible',
                backgroundColor: this.getColor()
              }}
            />
          )
        }
      >
        {children}
      </Button>
    );

    if (tooltipContent) {
      return (
        <Tooltip
          key={`curiosity-tooltip-${chartId}`}
          content={tooltipContent}
          position={TooltipPosition.top}
          distance={0}
          entryDelay={100}
          exitDelay={0}
        >
          {button}
        </Tooltip>
      );
    }

    return button;
  }
}

/**
 * Prop types.
 *
 * @type {{chartId: string, children: Node, isDisabled: boolean, tooltipContent: Node, chart: object,
 *    isThreshold: boolean, isToggled: boolean, getToggle: Function}}
 */
C3GraphCardLegendItem.propTypes = {
  chart: PropTypes.shape({
    color: PropTypes.func,
    focus: PropTypes.func,
    hide: PropTypes.func,
    revert: PropTypes.func,
    toggle: PropTypes.func
  }),
  chartId: PropTypes.string.isRequired,
  children: PropTypes.node,
  isDisabled: PropTypes.bool,
  isThreshold: PropTypes.bool,
  isToggled: PropTypes.bool,
  tooltipContent: PropTypes.node,
  getToggle: PropTypes.func
};

/**
 * Default props.
 *
 * @type {{children: null, isDisabled: boolean, tooltipContent: null, chart: {hide: Function,
 *    color: Function, focus: Function, revert: Function, toggle: Function}, isThreshold: boolean,
 *    isToggled: boolean, getToggle: Function}}
 */
C3GraphCardLegendItem.defaultProps = {
  chart: {
    color: helpers.noop,
    focus: helpers.noop,
    hide: helpers.noop,
    revert: helpers.noop,
    toggle: helpers.noop
  },
  children: null,
  isDisabled: false,
  isThreshold: false,
  isToggled: false,
  tooltipContent: null,
  getToggle: null
};

export { C3GraphCardLegendItem as default, C3GraphCardLegendItem };
