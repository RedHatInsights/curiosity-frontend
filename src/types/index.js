import { rhsmApiTypes } from './rhsmApiTypes';

const apiTypes = {
  rhsmApi: rhsmApiTypes
};

export { apiTypes as default, apiTypes, rhsmApiTypes };
