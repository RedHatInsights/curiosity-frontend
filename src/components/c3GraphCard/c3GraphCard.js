import React from 'react';
import PropTypes from 'prop-types';
import { Card, CardHead, CardActions, CardBody } from '@patternfly/react-core';
import { Skeleton, SkeletonSize } from '@redhat-cloud-services/frontend-components/components/Skeleton';
import _isEqual from 'lodash/isEqual';
import { Select } from '../form/select';
import { connectTranslate, reduxActions, reduxSelectors, reduxTypes, store } from '../../redux';
import { helpers, dateHelpers } from '../../common';
import { rhsmApiTypes, RHSM_API_QUERY_GRANULARITY_TYPES as GRANULARITY_TYPES } from '../../types/rhsmApiTypes';
import { c3GraphCardHelpers } from './c3GraphCardHelpers';
import { C3GraphCardLegendItem } from './c3GraphCardLegendItem';
import { graphCardTypes } from '../graphCard/graphCardTypes';
import { C3Chart } from '../c3Chart/c3Chart';

/**
 * A chart/graph card.
 *
 * @augments React.Component
 * @fires onUpdateGraphData
 * @fires onGranularitySelect
 */
class C3GraphCard extends React.Component {
  state = {};

  componentDidMount() {
    this.onUpdateGraphData();
  }

  componentDidUpdate(prevProps) {
    const { graphQuery, productId } = this.props;

    if (productId !== prevProps.productId || !_isEqual(graphQuery, prevProps.graphQuery)) {
      this.onUpdateGraphData();
    }
  }

  /**
   * Call the RHSM APIs, apply filters.
   *
   * @event onUpdateGraphData
   */
  onUpdateGraphData = () => {
    const { getGraphReportsCapacity, graphQuery, isDisabled, productId } = this.props;
    const graphGranularity = graphQuery && graphQuery[rhsmApiTypes.RHSM_API_QUERY_GRANULARITY];

    if (!isDisabled && graphGranularity && productId) {
      const { startDate, endDate } = dateHelpers.getRangedDateTime(graphGranularity);
      const query = {
        [rhsmApiTypes.RHSM_API_QUERY_START_DATE]: startDate.toISOString(),
        [rhsmApiTypes.RHSM_API_QUERY_END_DATE]: endDate.toISOString(),
        ...graphQuery
      };

      getGraphReportsCapacity(productId, query);
    }
  };

  /**
   * On granularity select, dispatch granularity type.
   *
   * @event onGranularitySelect
   * @param {object} event
   */
  onGranularitySelect = (event = {}) => {
    const { value } = event;
    const { viewId } = this.props;

    store.dispatch({
      type: reduxTypes.rhsm.SET_GRAPH_GRANULARITY_RHSM,
      viewId,
      [rhsmApiTypes.RHSM_API_QUERY_GRANULARITY]: value
    });
  };

  /**
   * Apply a custom legend.
   *
   * @param {object} options
   * @param {Function} options.chart
   * @param {Array} options.filteredData
   * @param {string} options.granularity
   * @param {Array} options.hiddenDataFacets
   * @returns {Array}
   */
  renderLegend({ chart, filteredData = [], granularity, hiddenDataFacets = [] }) {
    const { state } = this;
    const { productId, productShortLabel, t } = this.props;

    return filteredData.map(({ id }) => {
      const isThreshold = /^threshold/.test(id);
      const isDataHidden = hiddenDataFacets.includes(id);
      const tooltipContent = (
        <p>{(isThreshold && t(`curiosity-graph.thresholdLegendTooltip`)) || t(`curiosity-graph.${id}LegendTooltip`)}</p>
      );

      const getToggle = ({ isToggled }) => this.setState({ [`${productId}-${granularity}-${id}`]: isToggled });
      const isToggled = state[`${productId}-${granularity}-${id}`] || false;

      return (
        <C3GraphCardLegendItem
          key={`legend-${productId}-${granularity}-${id}`}
          chart={chart}
          chartId={id}
          tooltipContent={tooltipContent}
          isDisabled={isDataHidden}
          isThreshold={isThreshold}
          isToggled={isToggled}
          getToggle={getToggle}
        >
          {(isThreshold && t(`curiosity-graph.thresholdLabel`)) ||
            t(`curiosity-graph.${id}Label`, { product: productShortLabel })}
        </C3GraphCardLegendItem>
      );
    });
  }

  /**
   * Apply props to chart/graph.
   *
   * @returns {Node}
   */
  renderChart() {
    const { filterGraphData, graphData, graphQuery, selectOptionsType, productId, productShortLabel } = this.props;

    const graphGranularity = graphQuery && graphQuery[rhsmApiTypes.RHSM_API_QUERY_GRANULARITY];
    const { selected } = graphCardTypes.getGranularityOptions(selectOptionsType);
    const updatedGranularity = graphGranularity || selected;

    if (!graphData || !Object.values(graphData).length) {
      return null;
    }

    const filtered = [];

    if (filterGraphData.length) {
      filterGraphData.forEach(filteredValue => {
        if (graphData[filteredValue.id]) {
          filtered.push({ ...filteredValue, data: [...graphData[filteredValue.id]] });
        }
      });
    } else {
      Object.keys(graphData).forEach(id => {
        filtered.push({ id, data: [...graphData[id]] });
      });
    }

    const { configuration = {}, hiddenDataFacets = [] } = c3GraphCardHelpers.c3Config({
      data: filtered,
      granularity: updatedGranularity,
      productShortLabel
    });

    return (
      <C3Chart key={`chart-${productId}-${updatedGranularity}`} config={configuration}>
        {({ chart }) =>
          this.renderLegend({
            chart,
            filteredData: filtered,
            granularity: updatedGranularity,
            hiddenDataFacets
          })
        }
      </C3Chart>
    );
  }

  /**
   * Render a chart/graph card with chart/graph.
   *
   * @returns {Node}
   */
  render() {
    const { cardTitle, children, error, graphQuery, isDisabled, selectOptionsType, pending, t } = this.props;

    if (isDisabled) {
      return null;
    }

    const { options } = graphCardTypes.getGranularityOptions(selectOptionsType);
    const graphGranularity = graphQuery && graphQuery[rhsmApiTypes.RHSM_API_QUERY_GRANULARITY];

    return (
      <Card className="curiosity-usage-graph fadein">
        <CardHead>
          <h2>{cardTitle}</h2>
          <CardActions>
            {children}
            <Select
              aria-label={t('curiosity-graph.dropdownPlaceholder')}
              onSelect={this.onGranularitySelect}
              options={options}
              selectedOptions={graphGranularity}
              placeholder={t('curiosity-graph.dropdownPlaceholder')}
            />
          </CardActions>
        </CardHead>
        <CardBody>
          <div className={`curiosity-skeleton-container ${(error && 'blur') || ''}`}>
            {pending && (
              <React.Fragment>
                <Skeleton size={SkeletonSize.xs} />
                <Skeleton size={SkeletonSize.sm} />
                <Skeleton size={SkeletonSize.md} />
                <Skeleton size={SkeletonSize.lg} />
              </React.Fragment>
            )}
            {!pending && this.renderChart()}
          </div>
        </CardBody>
      </Card>
    );
  }
}

/**
 * Prop types.
 *
 * @type {{productId: string, pending: boolean, error: boolean, graphQuery: object, cardTitle: string,
 *     filterGraphData: Array, getGraphReportsCapacity: Function, productShortLabel: string, selectOptionsType: string,
 *     viewId: string, t: Function, children: Node, graphData: object, isDisabled: boolean}}
 */
C3GraphCard.propTypes = {
  cardTitle: PropTypes.string,
  children: PropTypes.node,
  error: PropTypes.bool,
  filterGraphData: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      color: PropTypes.string
    })
  ),
  getGraphReportsCapacity: PropTypes.func,
  graphData: PropTypes.object,
  graphQuery: PropTypes.shape({
    [rhsmApiTypes.RHSM_API_QUERY_GRANULARITY]: PropTypes.oneOf([...Object.values(GRANULARITY_TYPES)]).isRequired
  }).isRequired,
  isDisabled: PropTypes.bool,
  pending: PropTypes.bool,
  productId: PropTypes.string.isRequired,
  selectOptionsType: PropTypes.oneOf(['default']),
  t: PropTypes.func,
  productShortLabel: PropTypes.string,
  viewId: PropTypes.string
};

/**
 * Default props.
 *
 * @type {{getGraphReportsCapacity: Function, productShortLabel: string, selectOptionsType: string,
 *     viewId: string, t: Function, children: null, pending: boolean, graphData: object,
 *     isDisabled: boolean, error: boolean, cardTitle: null, filterGraphData: Array}}
 */
C3GraphCard.defaultProps = {
  cardTitle: null,
  children: null,
  error: false,
  filterGraphData: [],
  getGraphReportsCapacity: helpers.noop,
  graphData: {},
  isDisabled: helpers.UI_DISABLED_GRAPH,
  pending: false,
  selectOptionsType: 'default',
  t: helpers.noopTranslate,
  productShortLabel: '',
  viewId: 'graphCard'
};

/**
 * Create a selector from applied state, props.
 *
 * @type {Function}
 */
const makeMapStateToProps = reduxSelectors.graphCard.makeGraphCard();

/**
 * Apply actions to props.
 *
 * @param {Function} dispatch
 * @returns {object}
 */
const mapDispatchToProps = dispatch => ({
  getGraphReportsCapacity: (id, query) => dispatch(reduxActions.rhsm.getGraphReportsCapacity(id, query))
});

const ConnectedGraphCard = connectTranslate(makeMapStateToProps, mapDispatchToProps)(C3GraphCard);

export { ConnectedGraphCard as default, ConnectedGraphCard, C3GraphCard };
