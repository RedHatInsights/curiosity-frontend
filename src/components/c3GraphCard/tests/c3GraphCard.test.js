import React from 'react';
import { shallow } from 'enzyme';
import { C3Chart } from '../../c3Chart/c3Chart';
import { C3GraphCard } from '../c3GraphCard';
import { RHSM_API_QUERY_GRANULARITY_TYPES as GRANULARITY_TYPES, rhsmApiTypes } from '../../../types/rhsmApiTypes';

describe('C3GraphCard Component', () => {
  it('should render a non-connected component', () => {
    const props = {
      query: { [rhsmApiTypes.RHSM_API_QUERY_GRANULARITY]: GRANULARITY_TYPES.DAILY },
      productId: 'lorem'
    };
    const component = shallow(<C3GraphCard {...props} />);

    expect(component).toMatchSnapshot('non-connected');
  });

  it('should render multiple states', () => {
    const props = {
      query: { [rhsmApiTypes.RHSM_API_QUERY_GRANULARITY]: GRANULARITY_TYPES.DAILY },
      productId: 'lorem',
      graphData: {
        physicalSockets: [
          {
            date: new Date('2019-06-01T00:00:00Z'),
            y: 10,
            x: 0
          },
          {
            date: new Date('2019-06-08T00:00:00Z'),
            y: 12,
            x: 1
          },
          {
            date: new Date('2019-06-25T00:00:00Z'),
            y: 3,
            x: 2
          }
        ]
      }
    };

    const component = shallow(<C3GraphCard {...props} />);

    component.setProps({
      error: true
    });

    expect({
      chartBarData: component.find(C3Chart).prop('config').data
    }).toMatchSnapshot('error passes values');

    component.setProps({
      status: 403
    });

    expect(component).toMatchSnapshot('error with 403 status');

    component.setProps({
      status: 500
    });

    expect(component).toMatchSnapshot('error with 500 status');

    component.setProps({
      error: false,
      pending: true
    });

    expect(component).toMatchSnapshot('pending');

    component.setProps({
      error: false,
      pending: false,
      fulfilled: true
    });

    expect(component).toMatchSnapshot('fulfilled');
  });

  it('should filter result sets', () => {
    const props = {
      filterGraphData: [
        {
          id: 'loremIpsumSockets'
        },
        {
          id: 'thresholdLoremIpsumSockets'
        }
      ],
      query: { [rhsmApiTypes.RHSM_API_QUERY_GRANULARITY]: GRANULARITY_TYPES.DAILY },
      productId: 'lorem',
      graphData: {
        loremIpsumSockets: [
          {
            date: new Date('2019-06-01T00:00:00Z'),
            y: 10,
            x: 0
          },
          {
            date: new Date('2019-06-08T00:00:00Z'),
            y: 12,
            x: 1
          },
          {
            date: new Date('2019-06-25T00:00:00Z'),
            y: 3,
            x: 2
          }
        ],
        dolorSitSockets: [
          {
            date: new Date('2019-06-01T00:00:00Z'),
            y: 10,
            x: 0
          },
          {
            date: new Date('2019-06-08T00:00:00Z'),
            y: 12,
            x: 1
          },
          {
            date: new Date('2019-06-25T00:00:00Z'),
            y: 3,
            x: 2
          }
        ],
        thresholdLoremIpsumSockets: [
          {
            date: new Date('2019-06-01T00:00:00Z'),
            y: 10,
            x: 0
          },
          {
            date: new Date('2019-06-08T00:00:00Z'),
            y: 12,
            x: 1
          },
          {
            date: new Date('2019-06-25T00:00:00Z'),
            y: 3,
            x: 2
          }
        ]
      }
    };

    const component = shallow(<C3GraphCard {...props} />);
    expect(component.find(C3Chart).prop('config').data).toMatchSnapshot('filtered config');
  });

  it('should render a custom legend', () => {
    const props = {
      query: { [rhsmApiTypes.RHSM_API_QUERY_GRANULARITY]: GRANULARITY_TYPES.DAILY },
      productId: 'lorem'
    };

    const component = shallow(<C3GraphCard {...props} />);
    const componentInstance = component.instance();

    expect(componentInstance.renderLegend({})).toMatchSnapshot('empty legend');

    expect(
      componentInstance.renderLegend({
        filteredData: [{ id: 'lorem' }, { id: 'thresholdIpsum' }],
        granularity: 'dolor',
        hiddenDataFacets: []
      })
    ).toMatchSnapshot('legend');
  });

  it('should return an empty render when disabled', () => {
    const props = {
      query: { [rhsmApiTypes.RHSM_API_QUERY_GRANULARITY]: GRANULARITY_TYPES.DAILY },
      isDisabled: true,
      productId: 'lorem'
    };
    const component = shallow(<C3GraphCard {...props} />);

    expect(component).toMatchSnapshot('disabled component');
  });
});
