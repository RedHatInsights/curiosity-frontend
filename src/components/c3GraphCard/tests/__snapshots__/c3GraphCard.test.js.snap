// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`C3GraphCard Component should filter result sets: filtered config 1`] = `
Object {
  "colors": Object {
    "loremIpsumSockets": undefined,
    "thresholdLoremIpsumSockets": "#4cb140",
  },
  "columns": Array [
    Array [
      "x",
      "2019-06-01",
      "2019-06-08",
      "2019-06-25",
    ],
    Array [
      "loremIpsumSockets",
      10,
      12,
      3,
    ],
    Array [
      "thresholdLoremIpsumSockets",
      10,
      12,
      3,
    ],
  ],
  "groups": Array [
    Array [
      "loremIpsumSockets",
    ],
  ],
  "names": Object {
    "loremIpsumSockets": "t(curiosity-graph.loremIpsumSocketsLabel, [object Object])",
    "thresholdLoremIpsumSockets": "t(curiosity-graph.thresholdLabel)",
  },
  "types": Object {
    "loremIpsumSockets": "area-spline",
    "thresholdLoremIpsumSockets": "step",
  },
  "x": "x",
}
`;

exports[`C3GraphCard Component should render a custom legend: empty legend 1`] = `Array []`;

exports[`C3GraphCard Component should render a custom legend: legend 1`] = `
Array [
  <C3GraphCardLegendItem
    chart={
      Object {
        "color": [Function],
        "focus": [Function],
        "hide": [Function],
        "revert": [Function],
        "toggle": [Function],
      }
    }
    chartId="lorem"
    getToggle={[Function]}
    isDisabled={false}
    isThreshold={false}
    isToggled={false}
    tooltipContent={
      <p>
        t(curiosity-graph.loremLegendTooltip)
      </p>
    }
  >
    t(curiosity-graph.loremLabel, [object Object])
  </C3GraphCardLegendItem>,
  <C3GraphCardLegendItem
    chart={
      Object {
        "color": [Function],
        "focus": [Function],
        "hide": [Function],
        "revert": [Function],
        "toggle": [Function],
      }
    }
    chartId="thresholdIpsum"
    getToggle={[Function]}
    isDisabled={false}
    isThreshold={true}
    isToggled={false}
    tooltipContent={
      <p>
        t(curiosity-graph.thresholdLegendTooltip)
      </p>
    }
  >
    t(curiosity-graph.thresholdLabel)
  </C3GraphCardLegendItem>,
]
`;

exports[`C3GraphCard Component should render a non-connected component: non-connected 1`] = `
<Card
  className="curiosity-usage-graph"
>
  <CardHeader>
    <CardTitle>
      <Title
        headingLevel="h2"
        size="lg"
      />
    </CardTitle>
    <CardActions
      className=""
    >
      <Select
        aria-label="t(curiosity-graph.dropdownPlaceholder)"
        ariaLabel="Select option"
        className=""
        id="generatedid-"
        isDisabled={false}
        isToggleText={true}
        name={null}
        onSelect={[Function]}
        options={
          Array [
            Object {
              "selected": true,
              "title": "t(curiosity-graph.dropdownDaily)",
              "value": "daily",
            },
            Object {
              "title": "t(curiosity-graph.dropdownWeekly)",
              "value": "weekly",
            },
            Object {
              "title": "t(curiosity-graph.dropdownMonthly)",
              "value": "monthly",
            },
            Object {
              "title": "t(curiosity-graph.dropdownQuarterly)",
              "value": "quarterly",
            },
          ]
        }
        placeholder="t(curiosity-graph.dropdownPlaceholder)"
        selectedOptions="daily"
        toggleIcon={null}
        variant="single"
      />
    </CardActions>
  </CardHeader>
  <CardBody>
    <div
      className="fadein"
    />
  </CardBody>
</Card>
`;

exports[`C3GraphCard Component should render multiple states: error passes values 1`] = `
Object {
  "chartBarData": Object {
    "colors": Object {
      "physicalSockets": undefined,
    },
    "columns": Array [
      Array [
        "x",
        "2019-06-01",
        "2019-06-08",
        "2019-06-25",
      ],
      Array [
        "physicalSockets",
        10,
        12,
        3,
      ],
    ],
    "groups": Array [
      Array [
        "physicalSockets",
      ],
    ],
    "names": Object {
      "physicalSockets": "t(curiosity-graph.physicalSocketsLabel, [object Object])",
    },
    "types": Object {
      "physicalSockets": "area-spline",
    },
    "x": "x",
  },
}
`;

exports[`C3GraphCard Component should render multiple states: error with 403 status 1`] = `
<Card
  className="curiosity-usage-graph"
>
  <CardHeader>
    <CardTitle>
      <Title
        headingLevel="h2"
        size="lg"
      />
    </CardTitle>
    <CardActions
      className="blur"
    >
      <Select
        aria-label="t(curiosity-graph.dropdownPlaceholder)"
        ariaLabel="Select option"
        className=""
        id="generatedid-"
        isDisabled={false}
        isToggleText={true}
        name={null}
        onSelect={[Function]}
        options={
          Array [
            Object {
              "selected": true,
              "title": "t(curiosity-graph.dropdownDaily)",
              "value": "daily",
            },
            Object {
              "title": "t(curiosity-graph.dropdownWeekly)",
              "value": "weekly",
            },
            Object {
              "title": "t(curiosity-graph.dropdownMonthly)",
              "value": "monthly",
            },
            Object {
              "title": "t(curiosity-graph.dropdownQuarterly)",
              "value": "quarterly",
            },
          ]
        }
        placeholder="t(curiosity-graph.dropdownPlaceholder)"
        selectedOptions="daily"
        toggleIcon={null}
        variant="single"
      />
    </CardActions>
  </CardHeader>
  <CardBody>
    <div
      className="blur"
    >
      <C3Chart
        className={null}
        config={
          Object {
            "axis": Object {
              "x": Object {
                "padding": 0,
                "tick": Object {
                  "format": [Function],
                },
                "type": "timeseries",
              },
              "y": Object {
                "default": Array [
                  0,
                  50,
                ],
                "min": 0,
                "padding": Object {
                  "bottom": 0,
                },
                "tick": Object {
                  "format": [Function],
                  "outer": false,
                  "show": false,
                },
              },
            },
            "data": Object {
              "colors": Object {
                "physicalSockets": undefined,
              },
              "columns": Array [
                Array [
                  "x",
                  "2019-06-01",
                  "2019-06-08",
                  "2019-06-25",
                ],
                Array [
                  "physicalSockets",
                  10,
                  12,
                  3,
                ],
              ],
              "groups": Array [
                Array [
                  "physicalSockets",
                ],
              ],
              "names": Object {
                "physicalSockets": "t(curiosity-graph.physicalSocketsLabel, [object Object])",
              },
              "types": Object {
                "physicalSockets": "area-spline",
              },
              "x": "x",
            },
            "grid": Object {
              "y": Object {
                "show": true,
              },
            },
            "legend": Object {
              "show": false,
            },
            "padding": Object {
              "bottom": 10,
              "left": 40,
              "right": 40,
              "top": 10,
            },
            "point": Object {
              "show": false,
            },
            "spline": Object {
              "interpolation": Object {
                "type": "monotone",
              },
            },
            "tooltip": Object {
              "format": Object {
                "title": [Function],
                "value": [Function],
              },
              "order": [Function],
            },
            "unloadBeforeLoad": true,
          }
        }
        key="chart-lorem-daily"
        onComplete={[Function]}
        style={Object {}}
      >
        <Component />
      </C3Chart>
    </div>
  </CardBody>
</Card>
`;

exports[`C3GraphCard Component should render multiple states: error with 500 status 1`] = `
<Card
  className="curiosity-usage-graph"
>
  <CardHeader>
    <CardTitle>
      <Title
        headingLevel="h2"
        size="lg"
      />
    </CardTitle>
    <CardActions
      className="blur"
    >
      <Select
        aria-label="t(curiosity-graph.dropdownPlaceholder)"
        ariaLabel="Select option"
        className=""
        id="generatedid-"
        isDisabled={false}
        isToggleText={true}
        name={null}
        onSelect={[Function]}
        options={
          Array [
            Object {
              "selected": true,
              "title": "t(curiosity-graph.dropdownDaily)",
              "value": "daily",
            },
            Object {
              "title": "t(curiosity-graph.dropdownWeekly)",
              "value": "weekly",
            },
            Object {
              "title": "t(curiosity-graph.dropdownMonthly)",
              "value": "monthly",
            },
            Object {
              "title": "t(curiosity-graph.dropdownQuarterly)",
              "value": "quarterly",
            },
          ]
        }
        placeholder="t(curiosity-graph.dropdownPlaceholder)"
        selectedOptions="daily"
        toggleIcon={null}
        variant="single"
      />
    </CardActions>
  </CardHeader>
  <CardBody>
    <div
      className="blur"
    >
      <C3Chart
        className={null}
        config={
          Object {
            "axis": Object {
              "x": Object {
                "padding": 0,
                "tick": Object {
                  "format": [Function],
                },
                "type": "timeseries",
              },
              "y": Object {
                "default": Array [
                  0,
                  50,
                ],
                "min": 0,
                "padding": Object {
                  "bottom": 0,
                },
                "tick": Object {
                  "format": [Function],
                  "outer": false,
                  "show": false,
                },
              },
            },
            "data": Object {
              "colors": Object {
                "physicalSockets": undefined,
              },
              "columns": Array [
                Array [
                  "x",
                  "2019-06-01",
                  "2019-06-08",
                  "2019-06-25",
                ],
                Array [
                  "physicalSockets",
                  10,
                  12,
                  3,
                ],
              ],
              "groups": Array [
                Array [
                  "physicalSockets",
                ],
              ],
              "names": Object {
                "physicalSockets": "t(curiosity-graph.physicalSocketsLabel, [object Object])",
              },
              "types": Object {
                "physicalSockets": "area-spline",
              },
              "x": "x",
            },
            "grid": Object {
              "y": Object {
                "show": true,
              },
            },
            "legend": Object {
              "show": false,
            },
            "padding": Object {
              "bottom": 10,
              "left": 40,
              "right": 40,
              "top": 10,
            },
            "point": Object {
              "show": false,
            },
            "spline": Object {
              "interpolation": Object {
                "type": "monotone",
              },
            },
            "tooltip": Object {
              "format": Object {
                "title": [Function],
                "value": [Function],
              },
              "order": [Function],
            },
            "unloadBeforeLoad": true,
          }
        }
        key="chart-lorem-daily"
        onComplete={[Function]}
        style={Object {}}
      >
        <Component />
      </C3Chart>
    </div>
  </CardBody>
</Card>
`;

exports[`C3GraphCard Component should render multiple states: fulfilled 1`] = `
<Card
  className="curiosity-usage-graph"
>
  <CardHeader>
    <CardTitle>
      <Title
        headingLevel="h2"
        size="lg"
      />
    </CardTitle>
    <CardActions
      className=""
    >
      <Select
        aria-label="t(curiosity-graph.dropdownPlaceholder)"
        ariaLabel="Select option"
        className=""
        id="generatedid-"
        isDisabled={false}
        isToggleText={true}
        name={null}
        onSelect={[Function]}
        options={
          Array [
            Object {
              "selected": true,
              "title": "t(curiosity-graph.dropdownDaily)",
              "value": "daily",
            },
            Object {
              "title": "t(curiosity-graph.dropdownWeekly)",
              "value": "weekly",
            },
            Object {
              "title": "t(curiosity-graph.dropdownMonthly)",
              "value": "monthly",
            },
            Object {
              "title": "t(curiosity-graph.dropdownQuarterly)",
              "value": "quarterly",
            },
          ]
        }
        placeholder="t(curiosity-graph.dropdownPlaceholder)"
        selectedOptions="daily"
        toggleIcon={null}
        variant="single"
      />
    </CardActions>
  </CardHeader>
  <CardBody>
    <div
      className="fadein"
    >
      <C3Chart
        className={null}
        config={
          Object {
            "axis": Object {
              "x": Object {
                "padding": 0,
                "tick": Object {
                  "format": [Function],
                },
                "type": "timeseries",
              },
              "y": Object {
                "default": Array [
                  0,
                  50,
                ],
                "min": 0,
                "padding": Object {
                  "bottom": 0,
                },
                "tick": Object {
                  "format": [Function],
                  "outer": false,
                  "show": false,
                },
              },
            },
            "data": Object {
              "colors": Object {
                "physicalSockets": undefined,
              },
              "columns": Array [
                Array [
                  "x",
                  "2019-06-01",
                  "2019-06-08",
                  "2019-06-25",
                ],
                Array [
                  "physicalSockets",
                  10,
                  12,
                  3,
                ],
              ],
              "groups": Array [
                Array [
                  "physicalSockets",
                ],
              ],
              "names": Object {
                "physicalSockets": "t(curiosity-graph.physicalSocketsLabel, [object Object])",
              },
              "types": Object {
                "physicalSockets": "area-spline",
              },
              "x": "x",
            },
            "grid": Object {
              "y": Object {
                "show": true,
              },
            },
            "legend": Object {
              "show": false,
            },
            "padding": Object {
              "bottom": 10,
              "left": 40,
              "right": 40,
              "top": 10,
            },
            "point": Object {
              "show": false,
            },
            "spline": Object {
              "interpolation": Object {
                "type": "monotone",
              },
            },
            "tooltip": Object {
              "format": Object {
                "title": [Function],
                "value": [Function],
              },
              "order": [Function],
            },
            "unloadBeforeLoad": true,
          }
        }
        key="chart-lorem-daily"
        onComplete={[Function]}
        style={Object {}}
      >
        <Component />
      </C3Chart>
    </div>
  </CardBody>
</Card>
`;

exports[`C3GraphCard Component should render multiple states: pending 1`] = `
<Card
  className="curiosity-usage-graph"
>
  <CardHeader>
    <CardTitle>
      <Title
        headingLevel="h2"
        size="lg"
      />
    </CardTitle>
    <CardActions
      className=""
    >
      <Select
        aria-label="t(curiosity-graph.dropdownPlaceholder)"
        ariaLabel="Select option"
        className=""
        id="generatedid-"
        isDisabled={false}
        isToggleText={true}
        name={null}
        onSelect={[Function]}
        options={
          Array [
            Object {
              "selected": true,
              "title": "t(curiosity-graph.dropdownDaily)",
              "value": "daily",
            },
            Object {
              "title": "t(curiosity-graph.dropdownWeekly)",
              "value": "weekly",
            },
            Object {
              "title": "t(curiosity-graph.dropdownMonthly)",
              "value": "monthly",
            },
            Object {
              "title": "t(curiosity-graph.dropdownQuarterly)",
              "value": "quarterly",
            },
          ]
        }
        placeholder="t(curiosity-graph.dropdownPlaceholder)"
        selectedOptions="daily"
        toggleIcon={null}
        variant="single"
      />
    </CardActions>
  </CardHeader>
  <CardBody>
    <div
      className="fadein"
    >
      <Loader
        skeletonProps={
          Object {
            "size": "sm",
          }
        }
        tableProps={Object {}}
        variant="graph"
      />
    </div>
  </CardBody>
</Card>
`;

exports[`C3GraphCard Component should return an empty render when disabled: disabled component 1`] = `""`;
